src/analysis/souffle/tags.dl

```diff
@@ -0,0 +1,41 @@
+//-----------------------------------------------------------------------------
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//-----------------------------------------------------------------------------
+
+// Definition of the base tag types that are used by this dataflow analysis.
+
+#ifndef SRC_ANALYSIS_SOUFFLE_TAGS_DL_
+#define SRC_ANALYSIS_SOUFFLE_TAGS_DL_
+
+// A symbol representing a taint tag. SecrecyTags indicate that the data upon which they are
+// annotated potentially contain information that should be handled with some sort of special care
+// and should not be exposed to some outside group. The conservative approximation for SecrecyTags
+// is, when you are in doubt whether a particular datum should have this tag, annotate it. This
+// causes these tags to spread like taint.
+.type SecrecyTag <: symbol
```

Comment:
Actually, I don't think we should have separate types for `Secrecy` and `Integrity`. We just have a `Tag` type that would have an "integrity` and `secrecy` components that are captured via relations: `mayHaveSecrecyTag`, `mustHaveIntegrityTag`, etc. A same label like `product_id` could be used for capturing integrity and secrecy.

Reply:
Agreed. There should be one type of tag and we just need two different sets of relations for each. Both the confidentiality and integrity components of labels are sets, but they use different lattices (one is ordered by subset and the other is ordered by superset -- as a result the meet/join are also flipped). With the way IFC is usually used, a pretty common label would be to have the same set for both confidentiality and integrity (e.g. C = {Alice}, I = {Alice} ), and keeping tags in the same type allows this common label.